packages/comms/src/ecl/queryGraph.ts

```diff
@@ -0,0 +1,803 @@
+    subgraphs: Subgraph[] = [];
+    vertices: Vertex[] = [];
+    edges: Edge[] = [];
+
+    xgmml: string = "";
+
+    constructor() {
+        this.clear();
+    }
+
+    clear() {
+        this.xgmml = "";
+
+        this.idx = {};
+        this.subgraphs = [];
+        this.vertices = [];
+        this.edges = [];
+    }
+
+    load(xgmml: string) {
+        this.clear();
+        this.merge(xgmml);
+    }
+
+    merge(xgmml: string) {
+        this.xgmml = xgmml;
+        const parser = new DOMParser();
+        const dom = parser.parseFromString(xgmml, "text/xml");
+        this.walkDocument(dom.documentElement, "0");
+    }
+
+    isSubgraph(item: GraphItem): item is Subgraph {
+        return item instanceof Subgraph;
+    }
+
+    isVertex(item: GraphItem): item is Vertex {
+        return item instanceof Vertex;
+    }
+
+    isEdge(item: GraphItem): item is Edge {
+        return item instanceof Edge;
+    }
+
+    getGlobalType(item: QueryGraph | Subgraph | Vertex | Edge): GRAPH_TYPE {
+        if (item instanceof Vertex) {
+            return GRAPH_TYPE.VERTEX;
+        } else if (item instanceof Edge) {
+            return GRAPH_TYPE.EDGE;
+        } else if (item instanceof Subgraph) {
+            return GRAPH_TYPE.SUBGRAPH;
+        } else if (item instanceof QueryGraph) {
+            return GRAPH_TYPE.GRAPH;
+        }
+        return GRAPH_TYPE.UNKNOWN;
+    }
+
+    getGlobalTypeString(item: QueryGraph | Subgraph | Vertex | Edge): GRAPH_TYPE_STRING {
+        if (item instanceof Vertex) {
+            return GRAPH_TYPE_STRING.VERTEX;
+        } else if (item instanceof Edge) {
+            return GRAPH_TYPE_STRING.EDGE;
+        } else if (item instanceof Subgraph) {
+            return GRAPH_TYPE_STRING.SUBGRAPH;
+        } else if (item instanceof QueryGraph) {
+            return GRAPH_TYPE_STRING.GRAPH;
+        }
+        return GRAPH_TYPE_STRING.UNKNOWN;
+    }
+
+    getItem(docNode: HTMLElement, id: string): Subgraph | Vertex | Edge {
+        if (!this.idx[id]) {
+            switch (docNode.tagName) {
+                case "graph":
+                    const subgraph = new Subgraph(this, id);
+                    this.subgraphs.push(subgraph);
+                    this.idx[id] = subgraph;
+                    break;
+                case "node":
+                    const vertex = new Vertex(this, id);
+                    this.vertices.push(vertex);
+                    this.idx[id] = vertex;
+                    break;
+                case "edge":
+                    const edge = new Edge(this, id);
+                    this.edges.push(edge);
+                    this.idx[id] = edge;
+                    break;
+                default:
+                    console.warn("Graph.getItem - Unknown Node Type!");
+                    break;
+            }
+        }
+        const retVal = this.idx[id];
+        Array.from(docNode.attributes).forEach(attr => {
+            retVal[attr.name] = attr.value;
```

Comment:
It's nice to have these CodeQL annotations; seems like it's always stuff I don't think I'd notice. 

Could perhaps define the type of this.idx as a Map<string, Subgraph | Vertex | Edge> rather than an Object, and that might take care of this one and the others it pointed out below that manipulate the return of getItem?